include/pybind11/detail/descr.h

```diff
@@ -53,9 +53,16 @@ constexpr descr<N1 + N2, Ts1..., Ts2...> operator+(const descr<N1, Ts1...> &a, c
     return plus_impl(a, b, make_index_sequence<N1>(), make_index_sequence<N2>());
 }
 
+template <size_t N>
+constexpr descr<N - 1> const_str(char const(&text)[N]) { return descr<N - 1>(text); }
+constexpr descr<0> const_str(char const(&)[1]) { return {}; }
+
+// The "_" might be defined as a macro - don't define it if so.
+#ifndef _
```

Comment:
Do we should add a check to ensure people don't accidentally add the old syntax back into PyBind11 code.

Reply:
We could probably just add a define to a CMake job? `-D_=1` for example?

Reply:
@Skylion007 What kind of check did you have in mind?

The only thing I can think of is something like:
```
#ifndef PYBIND11_DO_NOT_DEFINE_UNDERSCORE
#ifndef _
...
#endif
#endif
```

Only loosely related, would this work?
```
#define _ const_str
```

Just for background / FYI:

I love this PR, although it will be a one or two day cleanup job for me to replace `_` with `const_str` throughout Google's code base (including imported 3rd party code). After I'm done I'd add `PYBIND11_DO_NOT_DEFINE_UNDERSCORE` to our central build rule, to ensure the underscore cannot be used anymore.

Reply:
Got it, thanks; sorry the history wasn't apparent to me.

A define similar to `PYBIND11_DO_NOT_DEFINE_UNDERSCORE` will be very important for my purposes, to guard against backsliding. I could patch it in when importing pybind11 from github, but I believe it will be useful in other environments, too, and it's just a couple lines extra in the code, and maybe one or two sentences in the changelog.

Reply:
Hi @henryiii, did you see my comment above? Does `PYBIND11_DO_NOT_DEFINE_UNDERSCORE` work for everybody? I'd be happy to add that here.

Reply:
I don't see the point of adding yet another configuration option when you can just define `-D_=1` (which is the original problem with this) and that works fine? You could even do:

```c++
#define _
#include <pybind11/pybind11.h>
#undef _
```

Reply:
Actually at 2nd thought, what I really want is an opt-in rather than an opt-out.
I'll just apply an automatic patch when importing from github.

About my other question, do we need to repeat the complicated `const_str` code? Is there a catch with this?
```
#ifndef _
#define _ const_str
#endif
```

I just tried it quickly locally and it seems to work.

Reply:
Because you should always avoid defines if at all possible. The problem here is that _other_ people are defining `_`; we are not, this was and is `pybind11::_`. We are trying to play nice with bad actors, not become one ourselves! :)

Reply:
In that case a comment would be useful, e.g.
```
// Repeating the const_str code to avoid introducing a #define.
```